libs/execution/include/hpx/parallel/executors/parallel_executor.hpp

```diff
@@ -93,10 +95,23 @@ namespace hpx { namespace parallel { namespace execution
 
         /// Create a new parallel executor
         HPX_CONSTEXPR explicit parallel_policy_executor(
-                Policy l = detail::get_default_policy<Policy>::call(),
-                std::size_t spread = 4, std::size_t tasks = std::size_t(-1))
-          : policy_(l), num_spread_(spread), num_tasks_(tasks)
-        {}
+            Policy l = detail::get_default_policy<Policy>::call(),
+            std::size_t spread = 4, std::size_t tasks = std::size_t(-1))
+          : policy_(l)
+          , num_spread_(spread)
+          , num_tasks_(tasks)
+        {
+            if (num_tasks_ == std::size_t(-1))
+            {
+                auto pool = threads::get_self_ptr() ?
+                    threads::get_self_id()
```

Comment:
Isn't `get_self_ptr()` already returning the pointer you need? I think there is no need to call `get_self_id()` afterwards (it accesses a thread_local, so might be a bit more costly than expected). Also, older gcc will complain about a constexpr constructor not having an empty body. I think you could wrap the `if()` into a static function to call in the member initializer:
```
    , num_tasks_(init_num_tasks(tasks))
```
or similar to work around this.

Reply:
Yes, on the `get_self_ptr`. What I have is unnecessary. And I'll try removing the constexpr from the constructor.